Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+    case .block(let b):
+      return b.description
+    case .functionDefinition(let f):
+      return f.description
+    case .if(let ifs):
+      return ifs.description
+    case .expression(let e):
+      return e.description
+    case .switch(let sw):
+      return sw.description
+    case .`for`(let loop):
+      return loop.description
+    case .break:
+      return "break"
+    case .continue:
+      return "continue"
+    case .noop:
+      return ""
+    case .inline(let s):
+      return s
+    }
+  }
+}
+
+public enum Type: String, CustomStringConvertible {
+  case bool
+  case u8
+  case s8
+  case u32
+  case s32
+  case u64
+  case s64
+  case u128
+  case s128
+  case u256
+  case s256
+  case any
+
+  public var description: String {
+    return self.rawValue
+  }
+}
+
+public typealias TypedIdentifierList = [(Identifier, Type)]
+
+private func render(typedIdentifierList: TypedIdentifierList) -> String {
+  return typedIdentifierList.map({ (ident, type) in
+    switch type {
+    case .any:
+      return "\(ident)"
+    default:
+      return "\(ident): \(type)"
+    }
+  }).joined(separator: ", ")
+}
+
+public struct FunctionDefinition: CustomStringConvertible {
+  public let identifier: Identifier
+  public let arguments: TypedIdentifierList
+  public let returns: TypedIdentifierList
+  public let body: Block
+
+  public init(identifier: Identifier,
+              arguments: TypedIdentifierList,
+              returns: TypedIdentifierList = [],
+              body: Block) {
+    self.identifier = identifier
+    self.arguments = arguments
+    self.returns = returns
+    self.body = body
+  }
+
+  public var description: String {
+    let args = render(typedIdentifierList: self.arguments)
+
+    var ret = ""
+    if !self.returns.isEmpty {
+      let retargs = render(typedIdentifierList: self.returns)
+      ret = "-> \(retargs)"
+    }
+
+    return "\(self.identifier)(\(args)) \(ret) \(self.body)"
+  }
+}
+
+public struct VariableDeclaration: CustomStringConvertible {
+  public let declarations: TypedIdentifierList
+  public let expression: Expression?
+
+  public init(_ declarations: TypedIdentifierList, _ expression: Expression? = nil) {
+    self.declarations = declarations
+    self.expression = expression
+  }
+
+  public var catchableSuccesses: [Expression] {
+    return expression?.catchableSuccesses ?? []
+  }
+
+  public var description: String {
+    let decls = render(typedIdentifierList: self.declarations)
+    if self.expression == nil {
+      return "let \(decls)"
+    }
+    return "let \(decls) := \(self.expression!.description)"
+  }
+}
+
+public typealias IdentifierList = [Identifier]
+
+public struct Assignment: CustomStringConvertible {
+  public let identifiers: IdentifierList
+  public let expression: Expression
+
+  public init(_ identifiers: IdentifierList, _ rhs: Expression) {
+    self.identifiers = identifiers
+    self.expression = rhs
+  }
+
+  public var catchableSuccesses: [Expression] {
```

Comment:
A lot of the declarations seem to have these – could you enforce the name & signature through a protocol? That may also make it easier if we group all of the AST elements which have catchable successes